File: CasualConsole/CasualConsoleCore/Templater/templater.js
const Templater = (() => {
    const registeredHelpers = {};
    const compile = template => {
        const handlers = [];
        let end = 0;
        let handler;
        while (end < template.length) {
            [handler, end] = getHandler(template, end);
            if (!handler)
                throw new Error("Unexpected end token");
            handlers.push(handler);
        }
        return (data, helpers = {}) => {
            const parts = [];
            const writer = x => parts.push(x);
            const contextData = {};
            contextData["$"] = data;
            const context = {
                get(key) {
                    return contextData[key] ?? helpers[key] ?? registeredHelpers[key] ?? data[key];
                },
                set(name, val) {
                    contextData[name] = val;
                }
            };
            handleMulti(writer, context, handlers);
            return parts.join("");
        };
    };
    const registerHelper = (name, f) => registeredHelpers[name] = f;
    const getHandler = (template, start) => {
        if (start == template.length) throw new Error("Expected {{end}} but not found");
        const i = template.indexOf("{{", start);
        if (i == start) {
            let [tokens, end] = getTokens(template, i + 2);
            const first = tokens[0];
            if (first === "if") {
                const ifExpr = getExpression(tokens, 1);
                let ifHandlers;
                [ifHandlers, end] = getBodyHandlers(template, end);
                let nextHandlerType = getHandlerType(template, end);
                if (nextHandlerType !== "else") {
                    const simpleHandler = (writer, context) => {
                        if (ifExpr(context))
                            handleMulti(writer, context, ifHandlers);
                    };
                    return [simpleHandler, end];
                }
                const elseIfHandlers = [];
                let elseHandlers = null;
                let elseTokens;
                while (true) {
                    [elseTokens, end] = getTokens(template, end + 2);
                    if (elseTokens.length == 1) {
                        [elseHandlers, end] = getBodyHandlers(template, end);
                        break;
                    }
                    else {
                        if (elseTokens[1] !== "if") throw new Error("Unexpected expression after else");
                        const elseIfExpr = getExpression(elseTokens, 2);
                        let elseIfInnerHandlers;
                        [elseIfInnerHandlers, end] = getBodyHandlers(template, end);
                        elseIfHandlers.push([elseIfExpr, elseIfInnerHandlers]);
                    }
                    nextHandlerType = getHandlerType(template, end);
                    if (nextHandlerType !== "else")
                        break;
                }
                const handler = (writer, context) => {
                    if (ifExpr(context)) {
                        handleMulti(writer, context, ifHandlers);
                        return;
                    }
                    for (const [elseIfExpr, elseIfHand] of elseIfHandlers) {
                        if (elseIfExpr(context)) {
                            handleMulti(writer, context, elseIfHand);
                            return;
                        }
                    }
                    if (elseHandlers) {
                        handleMulti(writer, context, elseHandlers);
                        return;
                    }
                };
                return [handler, end];
            }
            else if (first === "for") {
                let loopType = tokens[2];
                if (loopType !== "in" && loopType !== "range") {
                    const inIndex = tokens.findIndex(x => x === "in" || x === "range");
                    if (inIndex > 2) {
                        tokens = mergeTokens(tokens, inIndex);
                        loopType = tokens[2];
                    }
                    else
                        throw new Error("For loops have to be either 'in' or 'range' loops");
                }
                const [t1, t2] = tokens[1].split(",");
                const loopValuesExpr = getExpression(tokens, 3);
                let handlers;
                [handlers, end] = getBodyHandlers(template, end);
                const handler = loopType === "in"
                    ? (writer, context) => {
                        const loopValues = loopValuesExpr(context);
                        if (loopValues == null) throw new Error(`Value of '${tokens.slice(3).join(" ")}' was not iterable`);
                        let i = 0;
                        for (const val of loopValues) {
                            context.set(t1, val);
                            if (t2)
                                context.set(t2, i);
                            handleMulti(writer, context, handlers);
                            i++;
                        }
                    }
                    : (writer, context) => {
                        const obj = loopValuesExpr(context);
                        for (const key in obj) {
                            context.set(t1, key);
                            context.set(t2, obj[key]);
                            handleMulti(writer, context, handlers);
                        }
                    };
                return [handler, end];
            }
            else if (first === "end") {
                return [null, end];
            }
            else if (first === "else") {
                throw new Error("Unexpected else token");
            }
            else if (first === "set") {
                const varName = tokens[1];
                const expr = getExpression(tokens, 2);
                const handler = (writer, context) => context.set(varName, expr(context));
                return [handler, end];
            }
            else {
                const expr = getExpression(tokens, 0);
                const handler = (writer, context) => writer(htmlEncode(expr(context)));
                return [handler, end];
            }
        }
        else if (i < 0) {
            return [template.substring(start), template.length];
        }
        else {
            return [template.substring(start, i), i];
        }
    }
    const mergeTokens = (tokens, inIndex) => [tokens[0], tokens.slice(1, inIndex).join(""), ...tokens.slice(inIndex)];
    const handleMulti = (writer, context, handlers) => {
        for (const h of handlers)
            if (typeof h === "string")
                writer(h);
            else
                h(writer, context);
    }
    const getTokens = (template, i) => {
        const tokens = [];
        while (true) {
            while (template[i] === ' ')
                i++;
            if (template[i] === '}' && template[i + 1] === '}')
                return [tokens, i + 2];
            const start = i++;
            while (i < template.length && template[i] !== '}' && template[i] !== ' ')
                i++;
            const token = template.substring(start, i);
            if (!token)
                throw new Error("Tag not closed with }}")
            tokens.push(token);
        }
    }
    const getHandlerType = (template, start) => {
        while (start < template.length && /\s/.test(template[start])) // checking for whitespace
            start++;
        if (start + 1 < template.length && template[start] === '{' && template[start + 1] === '{') {
            start += 2;
            while (template[start] === ' ')
                start++;
            const tempStart = start++;
            while (start < template.length && template[start] !== '}' && template[start] !== ' ')
                start++;
            return template.substring(tempStart, start);
        }
        return null;
    }
    const getBodyHandlers = (template, end) => {
        const handlers = [];
        let handler;
        while (true) {
            [handler, end] = getHandler(template, end);
            if (!handler)
                break;
            handlers.push(handler);
        }
        return [handlers, end];
    }
    const getExpression = (tokens, start) => {
        if (tokens.length - start == 0 || tokens[start] === '.') {
            if (tokens.length == 0) throw new Error("Expression cannot be empty");
            else throw new Error(`Expression expected after: ${tokens.slice(0, start).join(" ")}`);
        }
        if (tokens.length - start == 1)
            return getTokenAsExpression(tokens[start]);
        const f = tokens[start];
        if (f.includes(".")) throw new Error(`Function name cannot contain a dot character: ${f}`);
        const argGroups = getArgGroups(tokens, start + 1);
        if (argGroups.length == 1) {
            const expr = argGroups[0];
            return context => {
                const func = getFunc(f, context);
                return func(expr(context));
            }
        }
        return context => {
            const func = getFunc(f, context);
            const args = argGroups.map(expr => expr(context));
            return func.apply(null, args);
        }
    }
    const getFunc = (f, context) => {
        const func = context.get(f);
        if (typeof func !== "function")
            throw new Error(`value of ${f} was not a function`);
        return func;
    }
    const getTokenAsExpression = token => {
        const expr = getTokenAsExpressionInner(token);
        return context => {
            try {
                return expr(context);
            } catch (e) {
                throw new Error(`Error while resolving: ${token}`);
            }
        };
    }
    const getTokenAsExpressionInner = token => {
        const subTokens = token.split(".");
        subTokens.forEach(x => { if (!x) throw new Error(`Invalid member access expression: ${token}`) });
        const t = subTokens[0];
        if (subTokens.length == 1) {
            return context => context.get(t);
        }
        if (subTokens.length == 2) {
            const k = subTokens[1];
            return context => context.get(t)[k];
        }
        if (subTokens.length == 3) {
            const k1 = subTokens[1];
            const k2 = subTokens[2];
            return context => context.get(t)[k1][k2];
        }
        return context => {
            let o = context.get(t);
            for (let i = 1; i < subTokens.length; i++)
                o = o[subTokens[i]];
            return o;
        }
    }
    const getArgGroups = (tokens, index) => {
        const exprs = [];
        while (index < tokens.length)
            exprs.push(getTokenAsExpression(tokens[index++]));
        return exprs;
    }
    const htmlEncode = s => {
        s = String(s ?? "");
        return /[&<>\'\"]/.test(s)
            ? s.replace(/&/g, '&amp;')
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;')
                .replace(/'/g, '&#39;')
                .replace(/"/g, '&#34;')
            : s;
    }
    return { compile, registerHelper };
})();
export default Templater;

File: CasualConsole/CasualConsoleCore/Templater/Templater.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web;

namespace CasualConsoleCore.Templater;

public class Templater
{
    public static Func<object, Dictionary<string, object>, string> CompileTemplate(string template)
    {
        var handlers = new List<object>();
        var end = 0;
        while (end < template.Length)
        {
            (var handler, end) = GetHandler(template, end);
            if (handler == null)
                throw new Exception();
            handlers.Add(handler);
        }
        return (data, helpers) =>
        {
            var stringBuilder = new StringBuilder();
            Action<string> writer = x => stringBuilder.Append(x);
            HandleMulti(writer, new RealContext(helpers, data), handlers);
            return stringBuilder.ToString();
        };
    }

    private static (object, int) GetHandler(string template, int start)
    {
        var i = template.IndexOf("{{", start);
        if (i == start)
        {
            var (tokens, end) = GetTokens(template, i + 2);
            var first = tokens[0];
            if (first == "if")
            {
                var ifExpr = GetExpression(tokens, 1);
                (var ifHandlers, end) = GetBodyHandlers(template, end);
                var nextHandlerType = GetHandlerType(template, end);
                if (nextHandlerType != "else")
                {
                    Action<Action<string>, Context> simpleHandler = (writer, context) =>
                    {
                        if (Truthy(ifExpr(context)))
                        {
                            HandleMulti(writer, context, ifHandlers);
                        }
                    };
                    return (simpleHandler, end);
                }

                var elseIfHandlers = new List<(Func<Context, object>, List<object>)>();
                List<object>? elseHandlers = null;
                while (true)
                {
                    (var elseTokens, end) = GetTokens(template, end + 2);
                    if (elseTokens.Count == 1)
                    {
                        (elseHandlers, end) = GetBodyHandlers(template, end);
                        break;
                    }
                    else
                    {
                        // Has else if
                        if (elseTokens[1] != "if")
                            throw new Exception();
                        var elseIfExpr = GetExpression(elseTokens, 2);

                        (var elseIfInnerHandlers, end) = GetBodyHandlers(template, end);
                        elseIfHandlers.Add((elseIfExpr, elseIfInnerHandlers));
                    }
                    nextHandlerType = GetHandlerType(template, end);
                    if (nextHandlerType != "else")
                        break;
                }
                Action<Action<string>, Context> handler = (writer, context) =>
                {
                    if (Truthy(ifExpr(context)))
                    {
                        HandleMulti(writer, context, ifHandlers);
                        return;
                    }
                    foreach (var (elseIfExpr, elseIfHandlers) in elseIfHandlers)
                    {
                        if (Truthy(elseIfExpr(context)))
                        {
                            HandleMulti(writer, context, elseIfHandlers);
                            return;
                        }
                    }
                    if (elseHandlers != null)
                    {
                        HandleMulti(writer, context, elseHandlers);
                        return;
                    }
                };
                return (handler, end);
            }
            else if (first == "for")
            {
                var loopVarName = tokens[1];
                if (tokens[2] != "in")
                    throw new Exception();
                var loopValuesExpr = GetExpression(tokens, 3);
                (var handlers, end) = GetBodyHandlers(template, end);
                Action<Action<string>, Context> handler = (writer, context) =>
                {
                    var loopValues = loopValuesExpr(context);
                    foreach (var val in (object[])loopValues)
                    {
                        context.Set(loopVarName, val);
                        HandleMulti(writer, context, handlers);
                    }
                };
                return (handler, end);
            }
            else if (first == "end" || first == "else")
            {
                return (null, end);
            }
            else
            {
                var expr = GetExpression(tokens, 0);
                Action<Action<string>, Context> handler = (writer, context) =>
                {
                    var value = expr(context);
                    writer(HttpUtility.HtmlEncode(value?.ToString()));
                };
                return (handler, end);
            }
        }
        else if (i < 0)
        {
            return (CheckString(template[start..]), template.Length);
        }
        else
        {
            return (CheckString(template[start..i]), i);
        }
    }
    private static string CheckString(string s)
    {
        if (s.Length == 0)
            throw new Exception();
        return s;
    }
    private static void HandleMulti(Action<string> writer, Context context, List<object> handlers)
    {
        foreach (var handler in handlers)
        {
            if (handler is string s)
            {
                writer(s);
            }
            else
            {
                var h = (Action<Action<string>, Context>)handler;
                h(writer, context);
            }
        }
    }
    private static (List<string>, int) GetTokens(string template, int i)
    {
        var tokens = new List<string>();
        while (true)
        {
            while (template[i] == ' ')
                i++;

            if (template[i] == '}' && template[i + 1] == '}')
                return (tokens, i + 2);
            if (template[i] == '.')
            {
                tokens.Add(".");
                i++;
                continue;
            }

            int start = i++;
            while (i < template.Length && template[i] != '.' && template[i] != '}' && template[i] != ' ')
                i++;
            var token = template[start..i];
            if (token.Length == 0)
                throw new Exception();
            tokens.Add(token);
        }
    }
    private static string? GetHandlerType(string template, int start)
    {
        while (start < template.Length && char.IsWhiteSpace(template[start]))
            start++;

        if (start + 1 < template.Length && template[start] == '{' && template[start + 1] == '{')
        {
            start += 2;
            while (template[start] == ' ')
                start++;
            int tempStart = start++;
            while (template[start] != '.' && template[start] != '}' && template[start] != ' ')
                start++;
            return template[tempStart..start];
        }
        return null;
    }
    private static (List<object>, int) GetBodyHandlers(string template, int end)
    {
        var handlers = new List<object>();
        while (true)
        {
            (var handler, end) = GetHandler(template, end);
            if (handler == null)
                break;
            handlers.Add(handler);
        }
        return (handlers, end);
    }
    private static Func<Context, object> GetExpression(List<string> tokens, int start)
    {
        if (tokens.Count - start == 0)
            throw new Exception();
        if (tokens.Count - start == 1)
        {
            // has one token
            var token = tokens[start];
            return GetTokenAsExpression(token);
        }

        // has two or more tokens
        if (tokens[start + 1] == ".")
        {
            var argGroups = GetArgGroups(tokens, start);
            if (argGroups.Count == 1)
                return argGroups[0];

            throw new Exception();
        }
        else
        {
            // function call
            var f = tokens[start];
            var argGroups = GetArgGroups(tokens, start + 1);
            if (argGroups.Count == 1)
            {
                var expr = argGroups[0];
                return context =>
                {
                    var funcVal = context.Get(f);
                    if (funcVal is Func<object, object> func1)
                    {
                        var val = expr(context);
                        return func1(val);
                    }
                    throw new Exception();
                };
            }
            else
            {
                return context =>
                {
                    var funcVal = context.Get(f);
                    if (funcVal is Func<object[], object> func1)
                    {
                        var args = argGroups.Select(expr => expr(context)).ToArray();
                        return func1(args);
                    }
                    throw new Exception();
                };
            }
        }
    }
    private static Func<Context, object> GetTokenAsExpression(string token)
    {
        return context => Call(context.Get(token));
    }
    private static Func<Context, object> GetMemberAccessExpression(List<string> tokens, int start, int end)
    {
        if (end - start == 1)
        {
            return GetTokenAsExpression(tokens[start]);
        }
        else if (end - start == 3)
        {
            var objName = tokens[start];
            var key = tokens[start + 2];
            return context =>
            {
                var obj = context.Get(objName);
                if (key == "length")
                {
                    if (obj is object[] arr)
                        return arr.Length;
                    else if (obj is string s)
                        return s.Length;
                }
                return Call(obj.GetType().GetProperty(key).GetValue(obj));
            };
        }
        throw new Exception();
    }
    static object Call(object value)
    {
        if (value == null)
            return null;
        var valueType = value.GetType();
        if (valueType.IsGenericType && valueType.GetGenericTypeDefinition() == typeof(Func<>))
        {
            if (value is Func<object> f)
                return f();
            else
                throw new Exception();
        }
        return value;
    }
    private static List<Func<Context, object>> GetArgGroups(List<string> tokens, int start)
    {
        var args = new List<Func<Context, object>>();

        while (true)
        {
            int end = start;
            if (tokens[end] == ".") throw new Exception();
            end++;
            while (end < tokens.Count && tokens[end] == ".")
            {
                end++;
                if (end < tokens.Count && tokens[end] == ".") throw new Exception();
                end++;
            }

            args.Add(GetMemberAccessExpression(tokens, start, end));
            if (end == tokens.Count)
            {
                return args;
            }
            else
            {
                if (start == end) throw new Exception();
                start = end;
            }
        }
    }
    private static bool Truthy(object o)
    {
        if (o is null) return false;
        if (o is false) return false;
        if (o is string s) return s != "";
        if (o is int i) return i != 0;
        if (o is double d) return d != 0;
        return true;
    }

    interface Context
    {
        object Get(string key);
        void Set(string key, object value);
    }
    class RealContext : Context
    {
        Dictionary<string, object> values = new();
        Dictionary<string, object> helpers;
        object global;

        public RealContext(Dictionary<string, object> helpers, object global)
        {
            this.helpers = helpers;
            this.global = global;

            values["$"] = global;
        }

        public object? Get(string key)
        {
            return values.GetValueOrDefault(key)
                ?? helpers.GetValueOrDefault(key)
                ?? global.GetType().GetProperty(key)?.GetValue(global);
        }
        public void Set(string key, object value)
        {
            values[key] = value;
        }
    }
}
